clippy_lints/src/types.rs

```diff
@@ -2239,20 +2239,19 @@ impl<'a, 'b, 'tcx: 'a + 'b> ImplicitHasherConstructorVisitor<'a, 'b, 'tcx> {
 
 impl<'a, 'b, 'tcx: 'a + 'b> Visitor<'tcx> for ImplicitHasherConstructorVisitor<'a, 'b, 'tcx> {
     fn visit_body(&mut self, body: &'tcx Body) {
+        let prev_body = self.body;
         self.body = self.cx.tcx.body_tables(body.id());
         walk_body(self, body);
+        self.body = prev_body;
     }
 
     fn visit_expr(&mut self, e: &'tcx Expr) {
         if_chain! {
             if let ExprKind::Call(ref fun, ref args) = e.node;
             if let ExprKind::Path(QPath::TypeRelative(ref ty, ref method)) = fun.node;
             if let TyKind::Path(QPath::Resolved(None, ref ty_path)) = ty.node;
+            if same_tys(self.cx, self.target.ty(), self.body.expr_ty(e));
```

Comment:
this doesn't quite do the same thing. we used to not walk the expression if this check failed. Not sure what for though. Maybe performance?

Reply:
Yeah this should only differ in performance, I'll revert this change